### PR TITLE
Added Theme Config-generation

### PR DESCRIPTION
**Pull request by Agreon**

### CODE REVIEW

File: lib/ExtensionConfigWatcher.js [hunk 1]
```diff
@@ -7,15 +7,16 @@ const EventEmitter = require('events')
...
 
 class ExtensionConfigWatcher extends EventEmitter {
-  constructor (options) {
+  constructor (options, watchFolder) {
     super()
-    this.extensionFolder = process.env.APP_PATH ? path.join(process.env.APP_PATH, EXTENSION_FOLDER) : EXTENSION_FOLDER
+    const extensionFolder = process.env.APP_PATH ? path.join(process.env.APP_PATH, EXTENSION_FOLDER) : EXTENSION_FOLDER
     this.options = options || {}
     this.options.ignoreInitial = true
+    this.watchFolder = watchFolder || extensionFolder
```
Comment (daddy-cool) [inline lib/ExtensionConfigWatcher.js]: this should always be this.watchFolder = extensionFolder

File: lib/ThemeConfigWatcher.js [hunk 2]
```diff
@@ -0,0 +1,14 @@
+const path = require('path')
+const ExtensionConfigWatcher = require('./ExtensionConfigWatcher')
+
+const THEME_FOLDER = 'themes'
+
+class ThemeConfigWatcher extends ExtensionConfigWatcher {
+  constructor (options) {
+    options = options || {}
+    const themeFolder = process.env.APP_PATH ? path.join(process.env.APP_PATH, THEME_FOLDER) : THEME_FOLDER
```
Comment (daddy-cool) [inline lib/ThemeConfigWatcher.js]: move this after super() instead and make it this.watchFolder = ... instead

File: lib/ExtensionConfigWatcher.js [hunk 3]
```diff
@@ -7,15 +7,16 @@ const EventEmitter = require('events')
 const EXTENSION_FOLDER = 'extensions'
 
 class ExtensionConfigWatcher extends EventEmitter {
-  constructor (options) {
+  constructor (options, watchFolder) {
```
Comment (daddy-cool) [inline lib/ExtensionConfigWatcher.js]: remove obsolete watchFolder param

File: lib/ThemeConfigWatcher.js [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+    super(options, themeFolder)
```
Comment (daddy-cool) [inline lib/ThemeConfigWatcher.js]: remove obsolete watchFolder param

File: lib/actions/BackendAction.js [hunk 5]
```diff
@@ -120,6 +127,11 @@ class BackendAction {
         fsEx.outputJsonSync(path.join('', config.path, 'frontend', 'config.json'), extConfig.frontend, {spaces: 2})
         logger.info(`Updated extension config for frontend ${config.file.id}`)
       }
+
+      if (!extConfig.frontend && !extConfig.backend) {
```
Comment (daddy-cool) [inline lib/actions/BackendAction.js]: what is extConfig.frontend if not the theme config?
I find it kind of dangerous that we assume that if the extConfig object doesnt include frontend or backend that it MUST be a theme

File: lib/actions/FrontendAction.js [hunk 6]
```diff
@@ -30,20 +35,26 @@ class FrontendAction {
...
     const dcClient = new DcHttpClient(UserSettings.getInstance())
+    const themeConfigWatcher = new ThemeConfigWatcher()
     const frontendSetup = new FrontendSetup(dcClient)
     switch (action) {
       default:
       case 'start': {
         const frontend = new FrontendProcess(options, frontendSetup)
         frontend.run()
+          .then(() => {
+            themeConfigWatcher.start(() => {
```
Comment (pvomhoff) [inline lib/actions/FrontendAction.js]: The watcher will not be stopped when a SIGINT comes in so the process never stops

File: lib/actions/FrontendAction.js [hunk 7]
```diff
@@ -60,6 +71,24 @@ class FrontendAction {
...
+      times: 5,
+      interval: (count) => 100 * Math.pow(2, count)
+    }, (acb) => {
+      dcClient.generateExtensionConfig(config.file, AppSettings.getInstance().getId(), (err, extConfig) => {
+        if (err) return acb(new Error(`Could not generate Config for '${config.file.id}'`))
+        return acb(null, extConfig)
+      })
+    }, (err, extConfig) => {
+      if (err) return
+      fsEx.outputJsonSync(path.join('', config.path, 'config', 'app.json'), extConfig, {spaces: 2})
```
Comment (pvomhoff) [inline lib/actions/FrontendAction.js]: Currently the app.json looks like:
```json
{
  "frontend": {
    "language": "en-us",
    "publicPath": "http://dgthqh2gka0ky.cloudfront.net/shop_30685/@shopgate/theme-gmd/0.1.21",
    "appId": "shop_30685",
    "logo": "https://data.shopgate.com/shops/24327/webapp@2x.png",
    "colors": {},
    "shopName": "sgxs-testshop-us1",
    "forgotPasswordUrl": null,
    "webCheckoutShopify": null,
    "features": {
      "showReviews": true
    }
  },
  "warnings": [
    "Using default ({}) for key 'colors'",
    "Using default (null) for key 'forgotPasswordUrl'",
    "Using default (null) for key 'webCheckoutShopify'"
  ]
}
```

But it should only contain the content of the frontend property

File: lib/actions/FrontendAction.js [hunk 8]
```diff
@@ -59,6 +70,32 @@ class FrontendAction {
         break
       }
     }
+
+    process.on('SIGINT', () => {
+      themeConfigWatcher.stop(() => {
+        logger.info('SDK connection closed')
+        process.exit(0)
```
Comment (pvomhoff) [inline lib/actions/FrontendAction.js]: The frontend SDK is showing an end screen after SIGINT. This is interrupted by this `process.exit(0)` Please remove it

**Comment by pvomhoff** [comment]
The watcher should print a message when the upload of the config started

**Comment by pvomhoff** [comment]
Please resolve merge conflicts

**Pull request merged**